consensus/src/experimental/commit_phase.rs

```diff
@@ -0,0 +1,333 @@
+                            counters::ERROR_COUNT.inc();
+                            error!(
+                                epoch = new_ledger_info.commit_info().epoch(),
+                                error = err.to_string(),
+                                "Error occurred in sending the commit vote message."
+                            )
+                        }).ok(); // note that this message will also reach the node itself
+
+                    self.local_queue.push_back(CacheItem::new(
+                        vecblock,
+                        ledger_info,
+                    ));
+                }
+                msg = self.commit_msg_receiver.select_next_some() => {
+                    match msg {
+                        ConsensusMsg::CommitVoteMsg(request) => {
+                            monitor!(
+                                "process_commit_vote",
+                                self.process_commit_vote(&*request).await.map_err(|err| {
+                                    counters::ERROR_COUNT.inc();
+                                    error!(
+                                        error = err.to_string(),
+                                        "Error in processing commit vote."
+                                    )
+                                }).ok()
+                            );
+                        }
+                        ConsensusMsg::CommitDecisionMsg(request) => {
+                            monitor!(
+                                "process_commit_decision",
+                                self.process_commit_decision(*request).await.map_err(|err| {
+                                    counters::ERROR_COUNT.inc();
+                                    error!(
+                                        error = err.to_string(),
+                                        "Error in processing commit decision."
+                                    )
+                                }).ok()
+                            );
+                        }
+                        _ => {}
+                    };
+                }
+            }
+            let mut_local_queue = &mut self.local_queue;
+            let mut_local_cache = &mut self.local_cache;
+            loop {
+                let front = mut_local_queue.front().unwrap();
+                let front_chash = front
+                    .ledger_info_sig
+                    .ledger_info()
+                    .commit_info()
+                    .executed_state_id();
+                match mut_local_cache.entry(front_chash) {
+                    Entry::Occupied(lio) => {
+                        // cancel out an item from local_cache and an item from local_queue
+                        // TODO: given that every accumulated signatures has passed verification,
+                        // we might just need to count the number of signatures.
+                        if lio.get().verify_signatures(&self.verifier).is_ok() {
```

Comment:
don't need verify here? (we should maintain the invariant that every inserted signature is valid), check_voting_power is probably enough

Reply:
Yeah, I wrote down that optimization. I didn't know there is a method check_voting_power.